### PR TITLE
Use the v6 field played_q to do a more direct blunder rescoring

### PR DESCRIPTION
**Pull request by Naphthalin**
Draft for how to use the new field `played_q` from v6 training data https://github.com/LeelaChessZero/lc0/pull/1468/ in rescoring:

- compare `best_q` and `played_q`
- if their difference is above a threshold, a proven win was missed, or a proven loss played without being forced, rescore all positions until that position
- rescore q, d and m (last one is still missing)

There is currently no code dealing with the special case of playing a longer/shorter proven line which gives inaccurate m but accurate q and d.

In general, the current code base is split between the final position where no consecutive position exists, and all other positions for which a consecutive position exists. I suppose this split isn't needed for the v6 rescoring following https://github.com/LeelaChessZero/lc0/issue/1308/ though I worked with it here.

### CODE REVIEW

File: src/selfplay/loop.cc [hunk 1]
```diff
@@ -964,7 +912,7 @@ void ProcessFile(const std::string& file, SyzygyTablebase* tablebase,
           }
         }
       }
-      if (deblunderEnabled) {
+      if (deblunderEnabled && fileContents.back().visits > 0) {
```
Comment (Tilps) [inline src/selfplay/loop.cc:918]: Add a comment about the visits check.

File: src/selfplay/loop.cc [hunk 2]
```diff
@@ -979,39 +927,23 @@ void ProcessFile(const std::string& file, SyzygyTablebase* tablebase,
...
-                deblunderingStarted = true;
-              }
-            }
+          auto& cur = fileContents[history.GetLength() - 1];
+          // A blunder is defined by the played move being worse than the
+          // best move by a defined threshold, missing a forced win, or
+          // playing into a proven loss without being forced.
+          if ((cur.best_q - cur.played_q >
+              deblunderQBlunderThreshold) ||
+              (cur.best_q > -1 && cur.played_q < 1 &&
```
Comment (Tilps) [inline src/selfplay/loop.cc:944]: This check is not 'strictly' true - the cur.best_q == 1 check could be strengthened by using the invariance_info best_q is a proof bit - played_q doesn't have a similar bit we can check - but maybe at least fixing the former wouldn't hurt.

**Comment by masterkni6** [comment]
Is the rescorer supposed to change result_q and result_d of the blunder move to best_q and best_d? I don't really understand the training all too well but doesn't that make it think the blunder is a good move?

File: src/selfplay/loop.cc (hunk 1 above)
Comment (Naphthalin) [reply src/selfplay/loop.cc:918]: Done.

File: src/selfplay/loop.cc (hunk 2 above)
Comment (Naphthalin) [reply src/selfplay/loop.cc:944]: so throw in an additional check that `best_q` is a proof (so together with the eval it's a proven win)?

**Comment by Naphthalin** [comment]
> Is the rescorer supposed to change result_q and result_d of the blunder move to best_q and best_d? I don't really understand the training all too well but doesn't that make it think the blunder is a good move?

There might be a small misconception on your side about the training: The Lc0 training trains on positions, not on played moves; the played move is only relevant for the game outcome. Changing `result_q`/`result_d` when a blunder occurs is the equivalent of a human "experimenting" in a position, i.e. "the current position is slightly better for white. This move looks interesting, let's see what happens even though I think it's a bad move." It doesn't make the blunder itself look good, but the position where the blunder happened because of temp exploration.
Does that make sense?

**Comment by Tilps** [comment]
> Is the rescorer supposed to change result_q and result_d of the blunder move to best_q and best_d? I don't really understand the training all too well but doesn't that make it think the blunder is a good move?

No - it avoids having the game outcome after the blunder, affect the training target before the blunder.
Nothing about that causes the policy of the bad move to increase, or for the training target after the blunder to be changed. (Unless there are multiple blunders in the game... but the logic still holds there.)
The main risk of this change is deblundering too often per game.  Unless there are runs of multiple positions between each blunder - this change almost becomes like training on a high q ratio.  Which is known to be a tiny bit weaker than what we already have.  The hypothesis is that there exists some happy medium where runs of training target are left unchanged far enough that learnings propagate back through the game effectively, while not also learning to expect blunders so much.

File: src/selfplay/loop.cc (hunk 2 above)
Comment (Tilps) [reply src/selfplay/loop.cc:944]: yes replace cur.best_q == 1 with something like ((cur.invariance_info & 8 != 0) && cur.best_q == 1)

Comment (Naphthalin) [reply src/selfplay/loop.cc:944]: Done.

**Review by Tilps — APPROVED** [review]

**Pull request merged**